### ISSUE TITLE
Crucible panics with `caller must change state from LiveRepair when requesting a stop` during stress testing.

### ISSUE THREAD

**Issue by leftwo** [oxidecomputer/crucible#1103]
@faithanalog reported the following: 
i ran 100 instances at the same time (in parallel) on madrid. each 1cpu, 1gig of ram, with my standard set of fio jobs. 93 of them had nvme errors (of varying description). 5 of them panicked (we are getting logs and core dumps for investigation of this). 2 of them succeeded running all the tests

propolis logs/core dumps from the 5 panicks are in /home/iliana/pylons-2024-01-21-001 on jeeves (which you should be able to access, xie set it 755 and i can read it)

The five instances that reported a panic were:
```
3153e243-e706-42ef-82f0-f96dbdd42577 - worker-19
9e636ea6-4285-41f5-8f1c-55202a9993b4 - worker-34
b36cad58-7662-4130-ad6a-fc49239c2e63 - worker-43
c3859f3e-1b21-406b-af21-b821f96c2534 - worker-50
7314b0a3-0d39-4480-9068-067ba494683a - worker-80
```

The panic messages looked like this:
```
thread 'tokio-runtime-worker' panicked at /home/build/.cargo/git/checkouts/crucible-f3b5bdecdc6486d6/e71b10d/upstairs/src/downstairs.rs:689:17:
caller must change state from LiveRepair when requesting a stop
```

**Comment by leftwo** [comment]
Looking at: `c3859f3e-1b21-406b-af21-b821f96c2534 - worker-50`

`omdb db instances` gives us:

```
ID                                   NAME      STATE   PROPOLIS_ID                          SLED_ID                              HOST_SERIAL 
c3859f3e-1b21-406b-af21-b821f96c2534 worker-50 running 8be04e29-7bc7-41f4-b4ac-f7a880bf642f 55157b83-4f9e-4e81-9171-6ad2786a88e4 BRM44220001 
```

This leads us to the log file:
`oxz_propolis-server_8be04e29-7bc7-41f4-b4ac-f7a880bf642f.system-illumos-propolis-server:default.log`

From that, we see we have kicked out a downstairs because too many jobs have built up:

```
04:31:01.353Z WARN propolis-server (crucible-36ab48d8-cefc-48b4-b0a1-f9fff604bfbd): downstairs failed, too many outstanding jobs 57001
04:31:01.353Z INFO propolis-server (crucible-36ab48d8-cefc-48b4-b0a1-f9fff604bfbd): [1] client skip 57001 in process jobs because fault
04:31:01.363Z INFO propolis-server (crucible-36ab48d8-cefc-48b4-b0a1-f9fff604bfbd): [1] changed 57001 jobs to fault skipped
04:31:01.701Z INFO propolis-server (crucible-36ab48d8-cefc-48b4-b0a1-f9fff604bfbd): ds_transition from Active to Faulted
04:31:01.701Z INFO propolis-server (crucible-36ab48d8-cefc-48b4-b0a1-f9fff604bfbd): [1] Transition from Active to Faulted
```

cleanup, then repair:
```
04:31:02.678Z WARN propolis-server (crucible-36ab48d8-cefc-48b4-b0a1-f9fff604bfbd): exiting client task is ignoring message Ruok
04:31:02.678Z INFO propolis-server (crucible-36ab48d8-cefc-48b4-b0a1-f9fff604bfbd): client task is exiting
04:31:03.598Z INFO propolis-server (crucible-36ab48d8-cefc-48b4-b0a1-f9fff604bfbd): connecting to [fd00:1122:3344:102::e]:19013
04:31:03.602Z INFO propolis-server (crucible-36ab48d8-cefc-48b4-b0a1-f9fff604bfbd): ds_connection connected
04:31:03.615Z INFO propolis-server (crucible-36ab48d8-cefc-48b4-b0a1-f9fff604bfbd): downstairs client at Some([fd00:1122:3344:102::e]:19013) has region UUID c199fecd-a875-49bc-aedb-4d5f8679f889
04:31:03.615Z INFO propolis-server (crucible-36ab48d8-cefc-48b4-b0a1-f9fff604bfbd): Returning UUID:c199fecd-a875-49bc-aedb-4d5f8679f889 matches
04:31:04.217Z INFO propolis-server (crucible-36ab48d8-cefc-48b4-b0a1-f9fff604bfbd): ds_transition from Faulted to LiveRepairReady
```

**Comment by leftwo** [comment]
Finding disks for instance c385 (worker-50):
```
omdb db disks list | egrep "NAME|c385"
NAME        ID                                   SIZE   STATE    ATTACHED_TO                          
worker-50   36ab48d8-cefc-48b4-b0a1-f9fff604bfbd 64 GiB attached c3859f3e-1b21-406b-af21-b821f96c2534 
```

Only one disk, that info is here:
```
root@oxz_switch1:~# omdb db disks info 36ab48d8-cefc-48b4-b0a1-f9fff604bfbd 
HOST_SERIAL DISK_NAME INSTANCE_NAME PROPOLIS_ZONE                                            VOLUME_ID                            DISK_STATE 
BRM44220001 worker-50 worker-50     oxz_propolis-server_8be04e29-7bc7-41f4-b4ac-f7a880bf642f d6dc51c2-11d4-4aad-bccb-2a0d7263c76d attached   
HOST_SERIAL REGION                               ZONE                                              PHYSICAL_DISK                        
BRM42220046 ac62f08b-9d62-466f-8ba2-ae173b09adcd oxz_crucible_13bd1f3a-a5c3-491d-adf4-d234571dd7d9 1f307dcb-a1d1-4e86-9d5b-a6ce7e11697b 
BRM42220004 c199fecd-a875-49bc-aedb-4d5f8679f889 oxz_crucible_29fdc49e-ac11-416b-924b-3b7101de4c9e a62269b7-b952-4719-baa9-87e72bf4330b 
BRM42220081 91b1ba7c-d1d9-4f04-ad5f-77f28e193fe0 oxz_crucible_4e1f6838-117e-494b-a4c4-1fda6f584465 06268004-a9d4-42d2-9241-2bdb4bbde6e7
```

**Comment by leftwo** [comment]
From the propolis log, client 1 for this propolis is:
```
{"msg":"downstairs client at Some([fd00:1122:3344:102::e]:19013) has region UUID c199fecd-a875-49bc-aedb-4d5f8679f889","v":0,"name":"propolis-server","level":30,"time":"2024-01-22T04:07:21.922714748Z","hostname":"oxz_propolis-server_8be04e29-7bc7-41f4-b4ac-f7a880bf642f","pid":11453,"client":"1","":"downstairs","session_id":"c11248de-346e-4694-a326-55fee1332f5a","component":"crucible-36ab48d8-cefc-48b4-b0a1-f9fff604bfbd"}
```

From the omdb output, this gives us:
```
BRM42220004 c199fecd-a875-49bc-aedb-4d5f8679f889 oxz_crucible_29fdc49e-ac11-416b-924b-3b7101de4c9e a62269b7-b952-4719-baa9-87e72bf4330b 
```

This downstairs also matches the downstairs IP:Port.  So, what does the downstairs log tell us?
 Looking for logs on BRM42220004 we find:
 ```
 root@BRM42220004:~# ls -l /pool/ext/*/crypt/debug/oxz_crucible_29fdc49e-ac11-416b-924b-3b7101de4c9e/oxide-crucible-downstairs*c199fecd*
-rw-r--r--   1 root     root        4509 Jan 22 04:15 /pool/ext/08b21261-f6ef-462b-b52c-6968de07905a/crypt/debug/oxz_crucible_29fdc49e-ac11-416b-924b-3b7101de4c9e/oxide-crucible-downstairs:downstairs-c199fecd-a875-49bc-aedb-4d5f8679f889.log.1705896441
-rw-r--r--   1 root     root      101165 Jan 22 08:15 /pool/ext/08b21261-f6ef-462b-b52c-6968de07905a/crypt/debug/oxz_crucible_29fdc49e-ac11-416b-924b-3b7101de4c9e/oxide-crucible-downstairs:downstairs-c199fecd-a875-49bc-aedb-4d5f8679f889.log.1705898964
-rw-r--r--   1 root     root           0 Jan 22 12:15 /pool/ext/08b21261-f6ef-462b-b52c-6968de07905a/crypt/debug/oxz_crucible_29fdc49e-ac11-416b-924b-3b7101de4c9e/oxide-crucible-downstairs:downstairs-c199fecd-a875-49bc-aedb-4d5f8679f889.log.1705911301
-rw-r--r--   1 root     root           0 Jan 22 16:16 /pool/ext/08b21261-f6ef-462b-b52c-6968de07905a/crypt/debug/oxz_crucible_29fdc49e-ac11-416b-924b-3b7101de4c9e/oxide-crucible-downstairs:downstairs-c199fecd-a875-49bc-aedb-4d5f8679f889.log.1705925700
```

The first log file just shows the initial connection.  In the second log file, we see the disconnect, and the follow up reconnect to the upstairs that will want to repair:
```
04:31:02.575Z ERRO crucible: connection ([fd00:1122:3344:101::1:17]:56988) Exits with error: Error reading from Upstairs: Connection reset by peer (os error 131)
04:31:03.604Z INFO crucible: accepted connection from [fd00:1122:3344:101::1:17]:57942
04:31:03.604Z INFO crucible: Connection request from 36ab48d8-cefc-48b4-b0a1-f9fff604bfbd with version 5
04:31:03.604Z INFO crucible: upstairs UpstairsConnection { upstairs_id: 36ab48d8-cefc-48b4-b0a1-f9fff604bfbd, session_id: c11248de-346e-4694-a326-55fee1332f5a, gen: 1 } connected, version 5
04:31:03.609Z WARN crucible: Attempting RW takeover from UpstairsConnection { upstairs_id: 36ab48d8-cefc-48b4-b0a1-f9fff604bfbd, session_id: c11248de-346e-4694-a326-55fee1332f5a, gen: 1 } to UpstairsConnection { upstairs_id: 36ab48d8-cefc-48b4-b0a1-f9fff604bfbd, session_id: c11248de-346e-4694-a326-55fee1332f5a, gen: 1 }
04:31:03.609Z WARN crucible: Signaling to UpstairsConnection { upstairs_id: 36ab48d8-cefc-48b4-b0a1-f9fff604bfbd, session_id: c11248de-346e-4694-a326-55fee1332f5a, gen: 1 } thread that UpstairsConnection { upstairs_id: 36ab48d8-cefc-48b4-b0a1-f9fff604bfbd, session_id: c11248de-346e-4694-a326-55fee1332f5a, gen: 1 } is being promoted (read-write)
04:31:03.609Z ERRO crucible: Error while signaling to UpstairsConnection { upstairs_id: 36ab48d8-cefc-48b4-b0a1-f9fff604bfbd, session_id: c11248de-346e-4694-a326-55fee1332f5a, gen: 1 } thread: SendError { .. }
04:31:03.609Z WARN crucible: Crucible Downstairs promoting UpstairsConnection { upstairs_id: 36ab48d8-cefc-48b4-b0a1-f9fff604bfbd, session_id: c11248de-346e-4694-a326-55fee1332f5a, gen: 1 } to active, discarding 13 jobs
04:31:03.609Z INFO crucible: UpstairsConnection { upstairs_id: 36ab48d8-cefc-48b4-b0a1-f9fff604bfbd, session_id: c11248de-346e-4694-a326-55fee1332f5a, gen: 1 } is now active (read-write)
04:31:04.214Z INFO crucible: Current flush_numbers [0..12]: [631, 162, 193, 224, 256, 289, 321, 353, 385, 419, 450, 482]
    task = proc
04:31:04.214Z INFO crucible: Downstairs has completed Negotiation
```

Some interesting info here from these logs, the downstairs knows this is a reconnection:
```
04:31:03.609Z WARN crucible: Attempting RW takeover from UpstairsConnection { upstairs_id: 36ab48d8-cefc-48b4-b0a1-f9fff604bfbd, session_id: c11248de-346e-4694-a326-55fee1332f5a, gen: 1 } to UpstairsConnection { upstairs_id: 36ab48d8-cefc-48b4-b0a1-f9fff604bfbd, session_id: c11248de-346e-4694-a326-55fee1332f5a, gen: 1 }
```
And, that the session IDs match.  
However, it is not able to communicate with the other downstairs task:
```
04:31:03.609Z ERRO crucible: Error while signaling to UpstairsConnection { upstairs_id: 36ab48d8-cefc-48b4-b0a1-f9fff604bfbd, session_id: c11248de-346e-4694-a326-55fee1332f5a, gen: 1 } thread: SendError { .. }
```

It makes me wonder if there is some deadlock in the downstairs.

Then the repair starts up:
```
04:32:30.282Z INFO crucible: Created copy dir "/data/regions/c199fecd-a875-49bc-aedb-4d5f8679f889/00/000/34E.copy"
04:32:30.567Z INFO crucible: eid:846 Found repair files: ["34E"]
04:32:36.010Z INFO crucible: Repair files downloaded, move directory "/data/regions/c199fecd-a875-49bc-aedb-4d5f8679f889/00/000/34E.copy" to "/data/regions/c199fecd-a875-49bc-aedb-4d5f8679f889/00/000/34E.replace"
04:32:36.011Z INFO crucible: Copy files from "/data/regions/c199fecd-a875-49bc-aedb-4d5f8679f889/00/000/34E.replace" in "/data/regions/c199fecd-a875-49bc-aedb-4d5f8679f889/00/000"
04:32:36.605Z INFO crucible: Move directory  "/data/regions/c199fecd-a875-49bc-aedb-4d5f8679f889/00/000/34E.replace" to "/data/regions/c199fecd-a875-49bc-aedb-4d5f8679f889/00/000/34E.completed"
04:32:36.944Z INFO crucible: Created copy dir "/data/regions/c199fecd-a875-49bc-aedb-4d5f8679f889/00/000/34F.copy"
04:32:36.999Z INFO crucible: eid:847 Found repair files: ["34F"]
```

**Comment by leftwo** [comment]
On to the repair failing, and the eventual panic:

On the propolis side, after repairing a bunch of extents, we get to 920, aka 0x398 we see this:
```
04:39:49.034Z INFO propolis-server (crucible-36ab48d8-cefc-48b4-b0a1-f9fff604bfbd): got repair ok in Closing { close_id: JobId(236138), repair_id: JobId(236139), noop_id: JobId(236140), reopen_id: JobId(236141), gw_repair_id: GuestWorkId(235140), gw_noop_id: GuestWorkId(235141) }
04:39:49.034Z INFO propolis-server (crucible-36ab48d8-cefc-48b4-b0a1-f9fff604bfbd): RE:920 Wait for result from repair command 236139:235140
04:39:49.034Z INFO propolis-server (crucible-36ab48d8-cefc-48b4-b0a1-f9fff604bfbd): Repair for extent 920 s:2 d:[ClientId(1)]
04:39:59.997Z INFO propolis-server (crucible-36ab48d8-cefc-48b4-b0a1-f9fff604bfbd): got repair ok in Repairing { repair_id: JobId(236139), noop_id: JobId(236140), reopen_id: JobId(236141), gw_noop_id: GuestWorkId(235141) }
04:39:59.997Z INFO propolis-server (crucible-36ab48d8-cefc-48b4-b0a1-f9fff604bfbd): RE:920 Wait for result from NoOp command 236140:235141
04:42:18.418Z INFO propolis-server (nvme-worker-50): Disabling controller
04:42:18.431Z INFO propolis-server (nvme-worker-50): Enabling controller
```

The downstairs says:
```
04:39:56.075Z INFO crucible: Created copy dir "/data/regions/c199fecd-a875-49bc-aedb-4d5f8679f889/00/000/398.copy"
04:39:56.127Z INFO crucible: eid:920 Found repair files: ["398"]
04:39:59.458Z INFO crucible: Repair files downloaded, move directory "/data/regions/c199fecd-a875-49bc-aedb-4d5f8679f889/00/000/398.copy" to "/data/regions/c199fecd-a875-49bc-aedb-4d5f8679f889/00/000/398.replace"
04:39:59.460Z INFO crucible: Copy files from "/data/regions/c199fecd-a875-49bc-aedb-4d5f8679f889/00/000/398.replace" in "/data/regions/c199fecd-a875-49bc-aedb-4d5f8679f889/00/000"
04:39:59.991Z INFO crucible: Move directory  "/data/regions/c199fecd-a875-49bc-aedb-4d5f8679f889/00/000/398.replace" to "/data/regions/c199fecd-a875-49bc-aedb-4d5f8679f889/00/000/398.completed"
04:49:24.034Z ERRO crucible: connection ([fd00:1122:3344:101::1:17]:57942) Exits with error: inactivity timeout
```

So, it looks like the downstairs is not taking action on the `NoOp`.  We don't know if it received it and was stuck, or tried to ack it and was stuck, or exactly what the state of the `NoOp` is.

**Comment by leftwo** [comment]
So, I think there are at least two issues here,

1) For the actual panic.  It looks like the panic happens if we try to stop the upstairs if we are in LiveRepair
2) Something happens that stops the downstairs from making progress.  First to build up enough jobs that then kick the downstairs out and require a repair, and then 2nd when trying to  do the LiveRepair.  This could be one issue or two.

**Comment by leftwo** [comment]
The panic should be now fixed with https://github.com/oxidecomputer/crucible/pull/1109, which fixes the timeout
not being noticed by moving he timeout check and stop request to come from inside the client task.

We still don't know what happend to the downstairs to prevent them from responding to the upstairs and that
might get moved to a new issue.

**Comment by leftwo** [comment]
The issue reported here could also be causing some problems: https://github.com/oxidecomputer/crucible/issues/1177
